File: Projeto1_Reginaldo_AEDII/src/RadixSort.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* Imprime um vetor de números inteiros, um elemento por linha. */
void print(int *vet, int N)
{
  int i;
  for (i = 0; i < N; i++)
    printf("%i\n", vet[i]);
}

/* Imprime um vetor de strings, um elemento por linha. */
void prints(char **nome, int N)
{
  int i;
  for (i = 0; i < N; i++)
  {
    puts(nome[i]);
  }
}

/* 
Calcula o índice não negativo em que a contagem do
caractere <letra> será armazenada no vetor de contagem
do algoritmo countingsort para vetores de strings.
Os caracteres são mapeados da seguinte maneira:
A-Z: números de 1 a 26
\0 e espaço: 0 (zero)
 */
int convert(char letra)
{
  int i;
  if (letra == ' ' || letra == '\0')
    return 0;
  if (letra >= 'A' && letra <= 'Z')
    i = (letra - 'A' + 1);
  return i;
}

/* Realiza o countingsort em um vetor de inteiros, considerando apenas
    o algarismo dado por p, onde dec = 10**p. */
void CountingSort(int *vetor, int N, int dec)
{
  int vet[10] = {0};
  int *ordenado = (int *)calloc(N, sizeof(int));
  int j, a;
  for (j = 0; j < N; j++)
  {
    a = (vetor[j] / dec) % 10;
    vet[a]++;
  }
  for (j = 0; j < 9; j++)
  {
    vet[j + 1] += vet[j];
  }
  for (j = N - 1; j >= 0; j--)
  {
    a = (vetor[j] / dec) % 10;
    ordenado[vet[a] - 1] = vetor[j];
    vet[a]--;
  }
  for (j = 0; j < N; j++)
    vetor[j] = ordenado[j];
  free(ordenado);
}

/* Realiza o countingsort em um vetor de strings, considerando apenas
    o caractere dado por p, onde p=0 representa o último caractere da
    string preenchida com espaços à sua direita até que seu tamanho seja
    igual a max, e p=max-1 é o primeiro caractere da string. */
void CountingSorts(char **nome, int N, int max, int p)
{
  int vet[27] = {0};
  char **ordenado = (char **)calloc(N, sizeof(char *));
  int j, c;
  for (j = 0; j < N; j++)
  {
    /*
    IMPORTANTE: Como as strings foram alocadas por calloc, elas já
    estão todas preenchidas com \0 até seus tamanhos serem iguais a max.
    Por conta disso, não é necessário calcular o verdadeiro índice
    dentro das strings, como nas implementações das outras linguagens de programação.
    */
    c = convert(nome[j][max - p - 1]);
    vet[c]++;
  }
  for (j = 0; j < 26; j++)
  {
    vet[j + 1] += vet[j];
  }
  for (j = N - 1; j >= 0; j--)
  {
    c = convert(nome[j][max - p - 1]);
    ordenado[vet[c] - 1] = nome[j];
    vet[c]--;
  }
  for (j = 0; j < N; j++)
    nome[j] = ordenado[j];
  free(ordenado);
}

/* Realiza o radixsort em um vetor de inteiros.
    max é o número de algarismos do maior número presente. */
void RadixSort(int *vet, int N, int max)
{
  int i, dec = 1;
  for (i = 0; i < max; i++)
  {
    CountingSort(vet, N, dec);
    dec *= 10;
  }
}

/* Realiza o radixsort em um vetor de strings.
    max é o número máximo de caracteres entre os elementos do vetor. */
void RadixSorts(char **nome, int N, int max)
{
  int i;
  for (i = 0; i < max; i++)
  {
    CountingSorts(nome, N, max, i);
  }
}

int main()
{
  char c[2], **nome;
  int i, *vet, N, max;
  /* Ver FORMATO DOS CASOS DE TESTE.txt para detalhes */
  scanf("%i", &N);
  scanf("%1s", c);
  scanf("%i", &max);
  if (c[0] == 'i')
  {
    vet = (int *)calloc(N, sizeof(int));
    for (i = 0; i < N; i++)
      scanf("%i", &vet[i]);
    RadixSort(vet, N, max);
    print(vet, N);
  }
  else if (c[0] == 's')
  {
    nome = (char **)calloc(N, sizeof(char *));
    getchar();
    for (i = 0; i < N; i++)
    {
      nome[i] = (char *)calloc(max + 2, sizeof(char));
      fgets(nome[i], max + 2, stdin);
      nome[i][strlen(nome[i]) - 1] = '\0';
    }
    RadixSorts(nome, N, max);
    prints(nome, N);
  }
  return 0;
}

File: Projeto1_Reginaldo_AEDII/bin/radix_sort.py
def prints(nome, N):
    '''
    Imprime a lista, um elemento por linha.
    '''
    for i in range(N):
        print(nome[i])


def countingSort(vet, exp1):
    '''
    Realiza o countingsort em um vetor de inteiros, considerando apenas
    o algarismo dado por a, onde exp1 = 10**a.
    '''
    n = len(vet)

    ordenado = [0] * (n)

    vet_cont = [0] * (10)

    for i in range(0, n):
        indice = (vet[i]/exp1)
        vet_cont[int((indice) % 10)] += 1

    for i in range(1, 10):
        vet_cont[i] += vet_cont[i-1]

    i = n-1
    while i >= 0:
        indice = (vet[i]/exp1)
        ordenado[vet_cont[int((indice) % 10)] - 1] = vet[i]
        vet_cont[int((indice) % 10)] -= 1
        i -= 1

    i = 0
    for i in range(0, len(vet)):
        vet[i] = ordenado[i]


def radixSort(vet, max_tam):
    '''
    Realiza o radixsort em um vetor de inteiros.
    max_tam é o número de algarismos do maior número presente.
    '''
    exp = 1
    while max_tam > 0:
        countingSort(vet, exp)
        exp *= 10
        max_tam -= 1

     # Para string:


def indice_char(string, max, i):
    '''
    Calcula o índice não negativo em que a contagem do
    caractere string[i] será armazenada no vetor de contagem
    do algoritmo countingsort para vetores de strings.
    i=0 representa o último caractere da string preenchida com espaços 
    à sua direita até que seu tamanho seja igual a max, e i=max-1 é o
    primeiro caractere da string. Os caracteres são mapeados da seguinte maneira:
    A-Z: números de 1 a 26
    Outros caracteres: 0 (zero)
    '''
    indice = 0
    dif_tam = max - len(string)

    if i >= dif_tam:
        indice_a_partir_do_final = i - dif_tam
        c = string[len(string) - 1 - indice_a_partir_do_final]
    else:
        c = ' '

    if c == ' ':
        indice = 0

    if 'A' <= c <= 'Z':
        indice = ord(c) - ord('A') + 1

    return indice


def ScountingSort(nome, N, max, i):
    '''
    Realiza o countingsort em um vetor de strings, considerando apenas
    o caractere dado por i, onde i=0 representa o último caractere da
    string preenchida com espaços à sua direita até que seu tamanho seja
    igual a max, e i=max-1 é o primeiro caractere da string.
    '''
    vet = [0] * 27
    ordenado = [None] * N
    for j in range(N):
        c = indice_char(nome[j], max, i)
        vet[c] += 1
    for j in range(1, 27):
        vet[j] += vet[j - 1]
    for j in range(N - 1, -1, -1):
        c = indice_char(nome[j], max, i)
        ordenado[vet[c] - 1] = nome[j]
        vet[c] -= 1
    for j in range(N):
        nome[j] = ordenado[j]
    a = 1


def SradixSort(nomes, N, max):
    '''
    Realiza o radixsort em um vetor de strings.
    max é o número máximo de caracteres entre os elementos do vetor.
    '''
    for i in range(max):
        ScountingSort(nomes, N, max, i)


def main():
    # Ver FORMATO DOS CASOS DE TESTE.txt para detalhes
    entrada = input()
    variaveis = entrada.split()
    N = variaveis[0]
    c = variaveis[1]
    max_tam = variaveis[2]

    N = int(N)
    max_tam = int(max_tam)

    if c == 'i':
        vet = [int(input()) for _ in range(N)]
        radixSort(vet, max_tam)
        prints(vet, N)
    else:

        nomes = []
        for _ in range(N):
            entrada = input().rstrip('\n')
            nomes.append(entrada)
        SradixSort(nomes, N, max_tam)
        prints(nomes, N)


if __name__ == "__main__":
    main()
